kcrypt/encryptor.go

```diff
@@ -0,0 +1,735 @@
+	err = formatLuks(device, info.Partition.Name, mapper, label, pass, e.logger)
+	if err != nil {
+		e.logger.Err(err).Msg("format luks")
+		return "", err
+	}
+
+	e.logger.Logger.Info().Str("label", label).Msg("Partition encryption completed")
+	return fmt.Sprintf("%s:%s:%s", info.Partition.FilesystemLabel, info.Partition.Name, info.Partition.UUID), nil
+}
+
+// getPasswordFromChallenger gets the password for a block.Partition using KcryptConfig.
+// It constructs the DiscoveryPasswordPayload internally for communication with the plugin.
+func (e *RemoteKMSEncryptor) getPasswordFromChallenger(b *block.Partition) (password string, err error) {
+	// Get a logger for debugging
+	log := types.NewKairosLogger("kcrypt-getPassword", "info", false)
+	defer log.Close()
+
+	log.Logger.Info().
+		Str("partition_name", b.Name).
+		Str("partition_label", b.FilesystemLabel).
+		Str("partition_uuid", b.UUID).
+		Msg("Requesting password for partition")
+
+	bus.Reload()
+
+	bus.Manager.Response(bus.EventDiscoveryPassword, func(_ *pluggable.Plugin, r *pluggable.EventResponse) {
+		password = r.Data
+		if r.Errored() {
+			err = fmt.Errorf("failed discovery: %s", r.Error)
+			log.Logger.Error().Err(err).Msg("Plugin returned error")
+		} else {
+			log.Logger.Info().
+				Int("password_length", len(password)).
+				Str("partition", b.Name).
+				Msg("DECRYPTION: Received password from plugin")
+		}
+	})
+
+	// Use kcryptConfig from the encryptor, scanning if not provided
+	kcryptConfig := e.kcryptConfig
+	if kcryptConfig == nil {
+		kcryptConfig = ScanKcryptConfig(e.logger)
+	}
+
+	// Construct DiscoveryPasswordPayload from KcryptConfig + partition
+	// This is where we create the payload that will be sent to kcrypt-challenger
+	var payload bus.DiscoveryPasswordPayload
+	payload.Partition = b
+	if kcryptConfig != nil {
+		payload.ChallengerServer = kcryptConfig.ChallengerServer
+		payload.MDNS = kcryptConfig.MDNS
+		log.Logger.Info().
+			Str("challenger_server", payload.ChallengerServer).
+			Msg("Using provided kcrypt config")
+	} else {
+		log.Logger.Info().Msg("No kcrypt config provided, using local encryption")
+	}
+
+	_, err = bus.Manager.Publish(bus.EventDiscoveryPassword, payload)
+	if err != nil {
+		log.Logger.Error().Err(err).Msg("Failed to publish event to bus")
+		return password, err
+	}
+
+	if password == "" {
+		log.Logger.Error().Msg("Received empty password from plugin")
+		return password, fmt.Errorf("received empty password")
+	}
+
+	log.Logger.Info().Msg("Password retrieval successful")
+	return
+}
+
+// TPMWithPCREncryptor encrypts partitions using TPM with PCR policy (UKI mode).
+type TPMWithPCREncryptor struct {
+	logger         types.KairosLogger
+	bindPublicPCRs []string
+	bindPCRs       []string
+}
+
+func (e *TPMWithPCREncryptor) Encrypt(partitions []string) error {
+	e.logger.Logger.Info().Str("method", e.Name()).Strs("partitions", partitions).Msg("Encrypting partitions")
+
+	_ = os.Setenv("SYSTEMD_LOG_LEVEL", "debug")
```

Comment:
Maybe we should move this to the luksifyMeasurements method directly. This is in case something goes wrong we have the most data vailable, not sure if it makes sense to have it here but rather just before and after we call the systemd commands

Reply:
Good point, moving.